File: app/src/test/java/me/cafecode/android/newspaper/NewsesPresenterTest.java
package me.cafecode.android.newspaper;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import me.cafecode.android.newspaper.data.LoadNewsesCallback;
import me.cafecode.android.newspaper.data.NewsRepository;
import me.cafecode.android.newspaper.data.models.News;
import me.cafecode.android.newspaper.newses.NewsesContract;
import me.cafecode.android.newspaper.newses.NewsesPresenter;

import static org.mockito.Mockito.verify;

/**
 * Created by Natthawut Hemathulin on 5/31/2017 AD.
 * Email: natthawut1991@gmail.com
 */

@RunWith(MockitoJUnitRunner.class)
public class NewsesPresenterTest {

    private NewsesPresenter mPresenter;

    @Mock
    NewsRepository mRepository;

    @Mock
    NewsesContract.View mView;

    @Captor
    ArgumentCaptor<LoadNewsesCallback> loadNewsesCallbackCaptor;

    @Before
    public void setUp() {
        mPresenter = new NewsesPresenter(mRepository, mView);
    }

    @Test
    public void onStart_whenPresenterStartWithoutCachedThenShowProgressView() {
        // Give

        // When
        mPresenter.onStart();

        // Then
        verify(mView).showProgressView(true);
    }

    @Test
    public void whenLoadNewsesFinishedThenShowNewses() {
        // Give
        List<News> newses = Arrays.asList(new News(), new News(), new News());

        // When
        mPresenter.loadNewses();

        verify(mView).showProgressView(true);

        // Callback
        verify(mRepository).loadNewses(loadNewsesCallbackCaptor.capture());
        loadNewsesCallbackCaptor.getValue().onLoadNewsesFinished(newses);

        // Then
        verify(mView).showProgressView(false);
        verify(mView).showNewses(newses);
    }
}

File: app/src/test/java/me/cafecode/android/newspaper/NewsesPresenterTest.kt
package me.cafecode.android.newspaper

import io.reactivex.Observable
import me.cafecode.android.newspaper.data.NewsRepository
import me.cafecode.android.newspaper.data.remote.NewsesResponse
import me.cafecode.android.newspaper.newses.NewsesContract
import me.cafecode.android.newspaper.newses.NewsesPresenter
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import utils.ReadJsonUtils
import io.reactivex.schedulers.Schedulers
import io.reactivex.android.plugins.RxAndroidPlugins
import me.cafecode.android.newspaper.util.ImmediateSchedulerProvider


@RunWith(MockitoJUnitRunner::class)
class NewsesPresenterTest {

    private val NEWSES_RESPONSE = ReadJsonUtils()
            .getJsonToMock("get_newses.json", NewsesResponse::class.java)

    lateinit var mPresenter: NewsesPresenter

    @Mock
    lateinit var mRepository: NewsRepository

    @Mock
    lateinit var mView: NewsesContract.View

    companion object {

        @BeforeClass
        @JvmStatic
        fun setUpClass() {
            // Override the default "out of the box" AndroidSchedulers.mainThread() Scheduler
            //
            // This is necessary here because otherwise if the static initialization block in AndroidSchedulers
            // is executed before this, then the Android SDK dependent version will be provided instead.
            //
            // This would cause a java.lang.ExceptionInInitializerError when running the test as a
            // Java JUnit test as any attempt to resolve the default underlying implementation of the
            // AndroidSchedulers.mainThread() will fail as it relies on unavailable Android dependencies.
            RxAndroidPlugins.setInitMainThreadSchedulerHandler { Schedulers.trampoline() }
        }

    }

    @Before
    fun setUp() {
        val scheduleProvider = ImmediateSchedulerProvider()
        mPresenter = NewsesPresenter(mRepository, mView, scheduleProvider)
    }

    @Test
    fun onStart_whenPresenterStartWithoutCachedThenShowProgressView() {
        // Give
        Mockito.`when`(mRepository.loadNewses())
                .thenReturn(Observable.just(NEWSES_RESPONSE.newses))

        // When
        mPresenter.onStart()

        // Then
        verify(mView)!!.showProgressView(true)
    }

    @Test
    fun whenLoadNewsesFinishedThenShowNewses() {
        // Give
        val newses = NEWSES_RESPONSE.newses
        Mockito.`when`(mRepository.loadNewses())
                .thenReturn(Observable.just(newses))

        // When
        mPresenter.loadNewses()

        // Then
        verify(mView)!!.showProgressView(true)

        // After load finish
        verify(mView)!!.showProgressView(false)
        verify(mView)!!.showNewses(newses)
    }

}
